FWCore/Framework/src/EventProcessor.cc

```diff
@@ -2131,25 +2155,53 @@ namespace edm {
 
   void EventProcessor::handleNextItemAfterMergingRunEntries(std::shared_ptr<RunProcessingStatus> iRunStatus,
                                                             WaitingTaskHolder iHolder) {
-    if (lastTransitionType() == InputSource::IsFile) {
-      iRunStatus->holderOfTaskInProcessRuns().doneWaiting(std::exception_ptr{});
-      iHolder.doneWaiting(std::exception_ptr{});
-    } else if (lastTransitionType() == InputSource::IsLumi && !iHolder.taskHasFailed()) {
-      CMS_SA_ALLOW try {
-        beginLumiAsync(IOVSyncValue(EventID(input_->run(), input_->luminosityBlock(), 0),
-                                    input_->luminosityBlockAuxiliary()->beginTime()),
-                       iRunStatus,
-                       iHolder);
-      } catch (...) {
-        WaitingTaskHolder copyHolder(iHolder);
-        iHolder.doneWaiting(std::current_exception());
-        endRunAsync(std::move(iRunStatus), std::move(iHolder));
+    chain::first([this, iRunStatus](auto nextTask) {
```

Comment:
The main action in the task in the ```first``` part of the chain is to call ```nextTransitionType```. Before the changes in this PR, this call always occurred in the function ```readAndMergeRunEntriesAsync```, which is contained in the ```first``` part of a different chain which contains the call to ```handleNextItemAfterMergingRunEntries``` in one of its ```then``` parts. It was always sequenced to call ```nextTransitionType``` before we reached this part of the code.

After the changes in this PR, there are two cases.

* In the first case the input source does not use the new feature declaring a run or lumi entry as the last to be merged. In this case, ```needToCallNext_``` is always false. The code in the ```first``` section does nothing. There is no benefit to having the chain in this case.
* In the other case, the Framework needs to know what is next before it knows whether to start a new lumi, end the run, start a new run, or pause for a file transition. It cannot proceed until ```nextTransitionType``` returns, because it doesn't know what to do yet. In addition, ```nextTransitionType``` interacts with the input source and may execute concurrently with other things interacting with the input source (for example, dereferencing a ```Ptr```). Therefore, we run it as a task in a serial queue (and protect it inside a mutex also). Because it is not running inside the same task and possibly on a different thread, we need the chain to time order completion of ```nextTransitionType``` and the transitions that follow (there are alternative ways to order this, but chain seemed the simplest alternative). That is the main benefit of the chain.

The chain also helps keep the sequence the same as before so that stream begin run executes before global begin lumi in the single threaded case, not really necessary but nice.

Reply:
Ah right, I see the need now. The lambda argument `nextTask` is being passed to `beginLumiAsync()` call.